File: h10/h10.rb
Pair = Struct.new(:fst,:snd)


def summingPairs(xs, sum)
  @the_pairs=[]  
  len = xs.length
  reader, writer = IO.pipe
  half = len/2
    threadA= Thread.fork do
        for i in 0..(len-1)
            for j in (i+1)..half
                if xs[i] + xs[j] <= sum
                    @the_pairs.push(Pair.new(xs[i],xs[j]))
                end 
            end
        end
        @the_pairs
    end
    threadB=Thread.fork do
        for i in 0..(len-1)
            for j in (half+1)..(len-1)
                if xs[i] + xs[j] <= sum
                    @the_pairs.push(Pair.new(xs[i],xs[j]))
                end 
            end
        end
        @the_pairs
    end
    threadA.join
    threadB.join
  return @the_pairs
end

File: h10/h10.clj


(defn summingPairs [xs sum]
    (letfn [(summingPairsHelper [xs ys the_pairs]
        (if (empty? xs) the_pairs
            (let [[fst & rest] xs]
              (recur
               rest ys
               (doall 
                (concat the_pairs
                        (for [snd ys 
                              :when (<= (+ fst snd) sum)]
                          [fst snd])))))))]
    ;; split the list into two sublist then use them concurrently
      (let [number (count xs)](
        let [left (quot number 2)
            right   (- number (quot number 2))]
            (let [left-item (take left xs)
                right-item (take-last right xs)]
                (let [list1  (future (summingPairsHelper xs left-item []))
                list2(future (summingPairsHelper xs right-item []))]
                (concat @list1 @list2)))))))





